### ISSUE TITLE
type checks

### ISSUE THREAD

**Issue by andre-merzky** [radical-cybertools/radical.entk#313]
about 10% to 20% of the EnTK code are type checks.  At the same time, that code is rather verbose and sometimes obfuscates the actual code semantics.

I would like to point out that SAGA also does  strong type checking on API level, but uses decorators for that.  As an example, see decorators [here](https://github.com/radical-cybertools/saga-python/blob/devel/src/saga/job/job.py) - search for `rus` (`radical.utils.signatures`).  FWIW, SAGA also typechecks return values this way.

I just wanted to point out this exists and is tested.  I should say though that the performance overhead of type checking in Python is in general not negligible, and that also holds for the decorator approach used in SAGA. - I disabled it thus for performance critical code paths.

**Comment by mturilli** [comment]
I would be happy to have a more homogeneous code base, with analogous implementation across systems. Said that, what would be the benefits of refactoring this code? Does it give any problem (bugs) and/or hinder further development? Not necessarily important but note that, so far, performance has not been an issue as EnTK overheads are consistently (much) smaller than other overheads involved in executing workflows at scale.

**Comment by andre-merzky** [comment]
I am going to ramble a bit, if you allow...

I think the type checking in RS is, technically, pretty cool - but the motivation was possibly not a good one.  SAGA is a well specified API, an I wanted to capture that in a language which, by design, is much less concerned about specificity.  Well - the approach works, and the checks are code wise not very intrusive.  But they do make the API slightly surprising **to Python users**.

Basically, we specify that a parameter can be mandatory or optional, and that it must be specified in one of a set of types.  We further ensure that a specific type is returned by a method.  

That means that, for example, a method like
```python
@rus.takes('File', 
           str,
           rus.optional(int, rus.nothing))
@rus.returns(int)
write(self, data, n=None):
    """
    write `n` bytes from `data` (all data if `n` is not specified),
    and return the number of bytes written
    """
    ....
```
would only accept a string and an optional integer.  That sounds good at face value - but:

  - why no unicode string?
  - why can't I write the **number** `3.1415926`?
  - why can `n` not be the float `1.0` which converts nicely to an `int`
  - why can `n` not be the string `"1"` which also converts on the fly?
  - why can `n` be `-1` instead, which makes no semantic sense?

Type constraints are either limiting and incomplete, or become quickly cumbersome, overly verbose, and are likely still limiting and incomplete.

What I would prefer nowadays is something like:

```python
write(self, data, n=None):
    try:
        if data: data = str(data)
        if n   : n    = int(n)
    except:
       raise TypeError('API cannot handle the given parameters')

    if n: 
        assert(n >= 0)
    if not data: 
        return  # nothing to do
```
So, I would attempt to convert whatever is passed by the user to the wanted type.  If that works, hurray.  If not, a `TypeError` is raised, which is exactly the error we want to raise.  Value checks are easy to do, and should be done anyway.

Its easy to convert that approach into some small convenient functions, like:
```
write(self, data, n=None):
    data = ru.assert_type(data, [str, None])
    n    = ru.assert_type(n, [int, None])
    n    = ru.assert_value("> 0", exclude=[None])
```
or whatever.  

Performance concerns remain - but relying on type conversion (`data = str(data)`) instead of type checking (`isinstance (data, str) or isinstance (data, unicode)`) is (a) faster, (b) more complete (`data = str(3,1415966)` is a string), (c) more concise, and (d) also actually converts instead of just checks - which one often needs anyway.

Be it as it may: I would only suggest to do type checking on API level, and only where it makes sense to check early.  Anything internal to a layer should be checked via unit tests or asserted, not type checked.  If we'd move to that level of detail, we should just switch language straight away.  I mean, there is really no point in re-inventing a type-safe compiler or interpreter...

EnTK's type checking is *very* verbose, in many methods it takes more code than the method semantics.  It is sometimes applied multiple times to the same data (`object.from_dict() checks the dict entries, and then, when setting members, the setters check types again).  EnTK is limiting in the types it accepts (e.g., no unicode for executable).  Most of this also holds for SAGA, FWIW.  RP does not do type checking (but does rely on some SAGA type checking for description types, and performs some value checking).

I am not actually suggesting any approach over the other.  I am not even suggesting to change anything, and would also accept to use EnTK's way everywhere (if a bit simplified via helpers), or to get rid of all type checking as most python modules do.  I just stumbled over this multiple times over the last days and wanted to see if people think this is worth discussing.

**Comment by mturilli** [comment]
Thanks Andre, this is interesting. @vivek-bala what is your take on this?

**Comment by vivek-bala** [comment]
> I would be happy to have a more homogeneous code base, with analogous implementation across systems. Said that, what would be the benefits of refactoring this code? Does it give any problem (bugs) and/or hinder further development?

I agree with Matteo here. I can see many benefits of it: ease of code reviews, all type and value checking functions/decorators can be checked once in RU (not in every tool RP/EnTK/tools using RP or EnTK), ease in maintaining and adding to this code base. Obviously, this would take time and there is possibility of errors during the conversion.

```python
    data = ru.assert_type(data, [str, None])
    n    = ru.assert_type(n, [int, None])
    n    = ru.assert_value("> 0", exclude=[None])
```

I think this is a good idea to try out (there might be more functions to add as well) and RU is the correct place for it.

> Be it as it may: I would only suggest to do type checking on API level, and only where it makes sense to check early. Anything internal to a layer should be checked via unit tests or asserted, not type checked.

Yes, there should be heavy type and value checking in all components that the user interacts with. As we discussed during the meeting, we want to capture these errors early. 

> EnTK's type checking is _very_ verbose, in many methods it takes more code than the method semantics. It is sometimes applied multiple times to the same data (`object.from_dict() checks the dict entries, and then, when setting members, the setters check types again). EnTK is limiting in the types it accepts (e.g., no unicode for executable).

Yes, I agree it does. But I would say it is worth it though. Regardless of the documentation we can provide (including API references), for objects that the user interacts with, I would definitely recommend it going forward as well. I do agree that between the type/value checking and the documentation, there might be some redundancy. 

More specifically, to_dict and from_dict can also be used by the users now. Initially they were private as they were only used for packaging and communication via rmq. The user can use them to print their workflow (using to_dict) or create their workflow (say based on a yaml file if they want). These two functions weren't originally created for the user (as I said only for communication), but I felt they can be quite userful for the user as well, hence made public (and hence heavily tested).

I don't think there are any points I disagree with really. Performance has not been a concern with our use cases so far. I think we need heavy type/value checking for user components and going with RU functions as proposed by Andre and using the same across all stacks (but testing them only in RU) sounds the correct thing to do.

**Comment by andre-merzky** [comment]
> to_dict and from_dict can also be used by the users now.

I need to correct myself on this one - tests are not **applied** twice, they are **defined** twice.    From task.py line 868:

```python
        if 'stdout' in d:
            if d['stdout']:
                if isinstance(d['stdout'], str) or isinstance(d['stdout'], unicode):
                    self._stdout = d['stdout']
                else:
                    raise TypeError(expected_type=str, actual_type=type(d['stdout']))
```
This should be 
```
self.stdout = d['stdout']
```

since the exact same type and value checks are in the setter for `self.stdout`.  Well they should be - they aren't actually, and that is a problem in itself ;-)  Same file, line 638:

```
    @stdout.setter
    def stdout(self, value):
        if isinstance(value, str):
            self._stdout = value
        else:
            raise TypeError(expected_type=str, actual_type=type(value))

```

**Comment by mturilli** [comment]
@andre-merzky if I understand your last message, you found a bug in the current implementation of the discussed approach? If so, would you mind to open a ticket about the bug?

**Comment by andre-merzky** [comment]
I don't think we need a new ticket.  Yes, the inconsistency can be considered a bug - but removing the code duplication as proposed in this ticket would also make the tests consistent (the same test is then used in both contexts).  Let me know if you want a ticket anyway (to have a different priority), and I'll open one or create a PR (this specific fix is simple and  localized).

**Comment by mturilli** [comment]
If the code is not broken (i.e., you found a case in which the program terminates with a trace), then I am fine with just this ticket.

**Comment by mturilli** [comment]
@andre-merzky @iparask time to pick this up again?

**Comment by andre-merzky** [comment]
Yes, I would be in favor of licking that up.  Compared to the discussion last time, I would add that RU now has a type safe dictionary class which does many of the things discussed here, and we should use it where appropriate (we use it for task descriptions and configs in RP and elsewhere).  I would not mind taking a stab at that on the RU level.

@iparask , @mtitov : would be glad to hear your input also.

**Comment by mtitov** [comment]
I would back up RU dictionary class (`Munch`) as Andre mentioned, I'm in favor of it and used it to describe some instances in RADICAL-DREAMER as well. It does define an instance structure and verifies attributes types [from the box]

**Comment by iparask** [comment]
I think it makes sense to use `ru.Munch` as @mtitov suggests. Let me try and create something for EnTK's task class and see how it works. I'll report as I move forward.

**Comment by mturilli** [comment]
@mtitov, @andre-merzky  was this ever done? If not, should we put it back on the development roadmap?

**Comment by andre-merzky** [comment]
This lives in fix/type_checks, but I am not sure about the current state.  From the looks of it it seems fairly complete. @mtitov ?
